lib/schema_generator/schema.rb

```diff
@@ -1,15 +1,65 @@
 # Read and write the schemas
 module SchemaGenerator
+  module SortedHash
+    refine Hash do
+      def deep_sort(&block)
+        recursive_deep_sort(self, &block)
+      end
+
+    private
+
+      def recursive_deep_sort(property, parent_key = nil, &block)
+        if property.is_a?(Hash)
+          children_sorted = property.each_with_object({}) do |(k, v), memo|
+            memo[k] = recursive_deep_sort(v, k, &block)
+          end
+
+          sorted = children_sorted.sort do |a, b|
+            block.call(a, b, parent_key)
+          end
+
+          Hash[sorted]
+        elsif property.respond_to?(:map)
+          property.map { |item| recursive_deep_sort(item, nil, &block) }
+        else
+          property
+        end
+      end
+    end
+  end
+
   class Schema
+    using SchemaGenerator::SortedHash
+
     def self.read(filename)
       @parsed_json_cache ||= {}
       @parsed_json_cache[filename] ||= JSON.parse(File.read(filename))
     end
 
     def self.write(full_filename, schema_hash)
-      schema_json = JSON.pretty_generate(schema_hash) + "\n"
+      schema_json = JSON.pretty_generate(ordered_schema(schema_hash)) + "\n"
       FileUtils.mkdir_p(File.dirname(full_filename))
       File.write(full_filename, schema_json)
     end
+
+    def self.ordered_schema(schema_hash)
+      # Custom consistent sorting for JSON Schema objects
+      sorted_hash = schema_hash.deep_sort do |(a), (b), parent_key|
```

Comment:
So when iterating the hash it iterates as arrays of keys and values.

eg:
```
[4] pry(main)> { k1: "v1", k2: "v2" }.map { |a| a }
=> [[:k1, "v1"], [:k2, "v2"]]
```
so options we have here to make it easier to understand are like:

```
sorted_hash = schema_hash.deep_sort do |(key_for_a, _), (key_for_b, _), parent_key|
```

or 
```
sorted_hash = schema_hash.deep_sort do |a, b, parent_key|
  key_for_a = a.first.to_s
  key_for_b = b.first.to_s
```

Not really sure if they help a whole lot though?